packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,224 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { joinUrlParts, stripTrailingSlash } from '../utils/url';
+
+/**
+ * Interface representing the structure of a route node in a nested object representation.
+ */
+export interface RouteNodeObjectRepresentation {
+  /** Optional metadata associated with this route node. */
+  metadata?: RouteTreeNodeMetadata;
+
+  /** Optional children nodes of this route node, keyed by their segment values. */
+  children?: Record<string, RouteNodeObjectRepresentation>;
+}
+
+/**
+ * Metadata for a route tree node.
+ */
+export interface RouteTreeNodeMetadata {
+  /** Optional redirect route associated with this node. */
+  redirectTo?: string;
+
+  /** The route segment for this node. */
+  route: string;
+}
+
+/**
+ * A node in a route tree.
+ */
+interface RouteTreeNode {
+  /** The value associated with this node (a route segment or wildcard). */
+  segment: string;
```

Comment:
Question: How is `segment` different from `metadata.route`?

Reply:
The `metadata.route` represents the full route up to the current node, provided it is a terminal route, while `segment` indicates the current route segment.

For example:

```json
{
  "segment": "en",
  "child": {
    "segment": "foo",
    "metadata": {
      "route": "en/foo"
    },
    "child": {
      "segment": "*",
      "metadata": {
        "route": "en/foo/:id"
      }
    }
  }
}
```

Reply:
Indeed, parameterized routes can get lost, so we could consider adding an additional field to store those parameters. However, using `metadata.route` is slightly more convenient since it eliminates the need to reconstruct the route path.